Produlytics/ui/front-end/src/app/model/http-interceptor/http-interceptor.service.ts

```diff
@@ -32,18 +32,33 @@ export class XhrInterceptor implements HttpInterceptor {
     req = req.clone({
       headers: req.headers.set('X-Requested-With', 'XMLHttpRequest')
     });
-    return next.handle(req)
-    .pipe(
-      catchError(
-        (error: HttpErrorResponse) => {
-          if (error.status === 401) {
-              this.loginService.logout();
 
-              return of();
-          }
+    /**
+     * Spring Boot di default è impostato in modo tale da bloccare le richieste REST che fanno del side effect
+     * (POST, PUT, DELETE, PATCH). Il modo per attivarle è mettere un header che si chiami X-XSFR-TOKEN e valorizzarlo
+     * con un cookie che ci dà il server. Questo è quanto si evince dalla documentazione di Spring Boot e da questo
+     * articolo https://www.baeldung.com/spring-security-csrf#2-front-end-configuration, da cui ho preso questa
+     * soluzione.
+     *
+     * Tutto ciò è parzialmente vero perché all'inizio il server non bloccava le POST, ma solo le PUT. Quindi potrebbe
+     * esserci dietro un bug di Spring Boot che non le impedisce. Resta il fatot che mettendo
+     * queste due linee di codice ora tutto funziona.
+     * */
+    const csrfToken = document.cookie.replace(/(?:(?:^|.*;\s*)XSRF-TOKEN\s*\=\s*([^;]*).*$)|^.*$/, '$1');
+    req.headers.set('X-XSRF-TOKEN', csrfToken);
```

Comment:
Ok, ma cosa fa esattamente quel `replace`? Senza quello mi sembra abbastanza chiaro che si stia prendendo un cookie e impostando un campo dell'header con il suo valore, ma perchè la regex strana?

Reply:
Devi estrarre il token dai cookie, che sono una big stringa enorme, in cui ci sta anche la roba che serve a noi

La regex l'ho copiata dall'articolo linkato e anche la "scelta" del replace, perché l'alternativa è fare un match -> group e tanto visto che replace non fa side effect tanto vale farlo così

Reply:
Ma non c'era tipo `CookieService` che ti permette di fare `cookieService.get("XSRF-TOKEN")`?

Poi cosa succede se quel cookie non è settato?

Reply:
Perché `CookieService` dava un errore con il cookie di sessione e ho supposto che l'avrebbe dato anche con con questo

Posso provare, anche se a naso dovrebbe funzionare visto che questo cookie non è httponly

Se il cookie non è settato allora c'è un errore grave col backend, perché come dicevo è configurazione standard di spring boot... Volendo posso mettere un controllo, se il cookie manca non lo metto negli header ma non credo che farà tanta differenza in una situazione irl